File: C++/problem-014/p14.cc
#include<iostream>
#include <chrono>
#include<math.h>
#include<string>
#include<array>
#include<fstream>
#include<algorithm>

//brute force, last 5 terms of sequence must be 1 2 4 8 16.
long int LengthChain(long int num);

int main(){
  std::chrono::time_point<std::chrono::system_clock> start, end;
  start = std::chrono::system_clock::now();
  // main program
  long int max = 5;
  long int specialnum = 16;
  for (long int i=17; i<1000000; i++){
    long int length = LengthChain(i);
    if (length>max){
      max = length;
      specialnum = i;
    }
  }
  std::cout << specialnum << std::endl;
  //end of main program
  end   = std::chrono::system_clock::now();
  std::chrono::duration<double> elapsed_seconds = end-start;
  std::cout << elapsed_seconds.count() << " seconds" <<std::endl;
}


long int LengthChain(long int num){
  int count = 0;
  while (num !=16){
    count++;
    if (num%2==0) num/=2;
    else num = num*3+1;
  }
  return count+5;
}

File: Python/problem-014/p14.py
import time

# last 5 terms are always 1 2 4 8 16

start = time.time()

def length(num):
    counter = 0
    while num!=16:
        counter = counter + 1
        if num%2 == 0: num = num/2
        else: num = num*3+1

    return(counter+5)

big = 5
specialnum = 16
for i in range(17,1000000):
    m = length(i)
    # print(m)
    if m > big:
        big = m
        specialnum = i

print(specialnum)
print(time.time() - start)
